pkg/plugin/clientmgmt/backupitemaction/v2/restartable_backup_item_action.go

```diff
@@ -0,0 +1,171 @@
+	return []AdaptedBackupItemAction{
+		{
+			Kind: common.PluginKindBackupItemActionV2,
+			GetRestartable: func(name string, restartableProcess process.RestartableProcess) biav2.BackupItemAction {
+				return NewRestartableBackupItemAction(name, restartableProcess)
+			},
+		},
+		{
+			Kind: common.PluginKindBackupItemAction,
+			GetRestartable: func(name string, restartableProcess process.RestartableProcess) biav2.BackupItemAction {
+				return NewAdaptedV1RestartableBackupItemAction(biav1cli.NewRestartableBackupItemAction(name, restartableProcess))
+			},
+		},
+	}
+}
+
+// restartableBackupItemAction is a backup item action for a given implementation (such as "pod"). It is associated with
+// a restartableProcess, which may be shared and used to run multiple plugins. At the beginning of each method
+// call, the restartableBackupItemAction asks its restartableProcess to restart itself if needed (e.g. if the
+// process terminated for any reason), then it proceeds with the actual call.
+type RestartableBackupItemAction struct {
+	Key                 process.KindAndName
+	SharedPluginProcess process.RestartableProcess
+}
+
+// NewRestartableBackupItemAction returns a new RestartableBackupItemAction.
+func NewRestartableBackupItemAction(name string, sharedPluginProcess process.RestartableProcess) *RestartableBackupItemAction {
+	r := &RestartableBackupItemAction{
+		Key:                 process.KindAndName{Kind: common.PluginKindBackupItemActionV2, Name: name},
+		SharedPluginProcess: sharedPluginProcess,
+	}
+	return r
+}
+
+// getBackupItemAction returns the backup item action for this restartableBackupItemAction. It does *not* restart the
+// plugin process.
+func (r *RestartableBackupItemAction) getBackupItemAction() (biav2.BackupItemAction, error) {
+	plugin, err := r.SharedPluginProcess.GetByKindAndName(r.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	backupItemAction, ok := plugin.(biav2.BackupItemAction)
+	if !ok {
+		return nil, errors.Errorf("%T (returned for %v) is not a BackupItemActionV2!", plugin, r.Key)
+	}
+
+	return backupItemAction, nil
+}
+
+// getDelegate restarts the plugin process (if needed) and returns the backup item action for this restartableBackupItemAction.
+func (r *RestartableBackupItemAction) getDelegate() (biav2.BackupItemAction, error) {
+	if err := r.SharedPluginProcess.ResetIfNeeded(); err != nil {
+		return nil, err
+	}
+
+	return r.getBackupItemAction()
+}
+
+// AppliesTo restarts the plugin's process if needed, then delegates the call.
+func (r *RestartableBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	delegate, err := r.getDelegate()
+	if err != nil {
+		return velero.ResourceSelector{}, err
+	}
+
+	return delegate.AppliesTo()
+}
+
+// Execute restarts the plugin's process if needed, then delegates the call.
+func (r *RestartableBackupItemAction) Execute(item runtime.Unstructured, backup *api.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, string, error) {
```

Comment:
Thinking of some plugin process restart cases:
1. If the plugin process quits unexpectedly (i.e. crash) during a method calling, the call will fail on the client side, so the backup/restore will fail because of the BIA/RIA method fails. Is that right?
2. Since the plugin runs some async operations, the plugin may quit outside a method calling, for example, the `Execute` method returns successfully but during running the async operation afterwards, the plugin crashes. In this case, I think the client side gets no notification from the plugin framework, but the specific BIA/RIA implementation needs to tell the caller during the later method calling (i.e., `Progress` method) if it wants to fail the current backup/restore. How do we cover this scenario?

Reply:
@Lyndon-Li
1. It should result in a partial failure, as the GRPC call itself will error out, but that's not any different from what happens with v1 plugins
2. If the plugin crashes without restarting, then the Progress method will return an error (similar to the Execute error in your first question), so this will push us to partial failure. If the plugin crashes but restarts, then it's up to the plugin to handle this. For example, for a datamover plugin which creates a separate datamover CR (where the actual async operation happens in a separate controller), a plugin restart after creating the datamover CR won't affect progress, as long as the plugin is running again by the time velero calls Progress next, since it just checks the datamover CR status. If the plugin is itself running the operation in a background thread of the plugin process, then a crash will probably result in an operation failure, so Progress will return an error.

Reply:
Sounds good, either way is allowed, the behavior is up to the specific plugin.